### PR TITLE
Add opt-in B2B wholesale portal

### PR DESCRIPTION
**Pull request by damianlegawiec**
## What

A gated **B2B wholesale portal** at `/wholesale` — a login-gated trade surface that shares one storefront app and Spree backend with the public DTC store. Approved buyers browse the catalog at channel-scoped trade pricing, place bulk orders via SKU quick-order, and check out on the wholesale channel; guests and DTC shoppers never see it.

Demonstrates blended DTC + B2B in a single storefront — the pattern none of the alternatives ship out of the box.

## Opt-in addon (off by default)

The portal turns on **only** when `SPREE_WHOLESALE_CHANNEL` names a gated channel on the backend. When unset:
- The `/wholesale` routes `404` (guarded in the route-group layout).
- The nav link, footer link, and homepage trade section are hidden.
- The storefront runs DTC-only with no wholesale code paths.

`SPREE_WHOLESALE_PUBLISHABLE_KEY` is optional — the `X-Spree-Channel` header selects the channel, so it falls back to `SPREE_PUBLISHABLE_KEY`; set it only to bind a channel-scoped key.

## How it works

- **Surfaces** (`dtc` | `wholesale`) select the SDK client, cart cookies, and cache keys. A shopper can hold independent DTC and wholesale carts at once while sharing one session (JWT). The shared checkout resolves the surface from the cart cookie.
- **Gating** is enforced server-side by the channel's `login_required` access (guests get `401` from the Store API — the storefront can't bypass it). **Approval** is Wholesale customer-group membership, read from `customer_groups` on the customer.
- **Trade pricing** unlocks by volume (10+ of a single item), surfaced on the PDP, cart nudge, and catalog note.

## Notable correctness work

Adapts the data layer to the SDK's **nullable money fields** (prices are `null` for prices-hidden guests). Guards added so a null amount is never coerced to zero — most importantly, **a null cart total can no longer complete checkout as a free order** — and a null price can't render a false discount. The post-login redirect is validated against same-origin paths (no open redirect).

## Requires

`@spree/sdk >= 1.2.0` (package.json bumped) — for the channel surface: `client.channel.get()`, `Cart.channel_id`, `customer_groups` on `customers/me`.

## Testing

- `tsc --noEmit`: clean · `biome check`: clean · `vitest`: 115/115
- Verified live against a seeded Spree backend: gate → sign-in → wholesale catalog at trade pricing → PDP → quick-order → cart → checkout on the wholesale channel; cross-surface cart isolation; addon disabled → routes 404 and entry points hidden.
- CodeRabbit reviewed; findings fixed (nullable-price guards, open-redirect, empty-key fallback) or skipped with rationale.

Docs: wholesale + channel-switching developer page added in `spree/spree` (`docs/developer/storefront/nextjs/wholesale.mdx`).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Touches cart, checkout, and payment amount logic across surfaces; misconfigured channel env or surface resolution could affect checkout routing, though guards and tests target cross-cart poisoning and null-price edge cases.
> 
> **Overview**
> Introduces an **opt-in wholesale B2B portal** at `/wholesale`, enabled only when `SPREE_WHOLESALE_CHANNEL` is set; otherwise routes 404 and nav/footer/home entry points stay hidden.
> 
> The portal runs as a second **surface** (`dtc` | `wholesale`) with its own Spree channel client, cart cookies, and cache tags so DTC and wholesale carts can coexist on one session. **WholesaleGate** branches on channel `storefront_access` (sign-in wall vs guest browse with hidden prices), Wholesale-group **approval**, and approved buyers get catalog/PDP, wholesale cart, SKU **quick-order**, and **apply** flows. Shared checkout resolves surface from the cart (with channel verification on offsite return).
> 
> **Cart/checkout hardening** drops poisoned cross-surface cart cookies, validates `channel_id`, clears all surface carts on logout, and treats nullable SDK money fields safely (null totals no longer read as free checkout; hidden-price prompts on gated catalog).
> 
> Storefront wiring adds homepage **WholesaleSection**, i18n for five locales, `HiddenPricingContext`, and bumps **`@spree/sdk` to ^1.2.0**. Docs/env/README describe wholesale env vars.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 573cfbc026263657f7c6b43dc33670175626c3aa. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

- **New Features**
  - Added an opt-in Wholesale B2B portal with wholesale navigation, gated sign-in, application submission, and localized pending/approved experiences.
  - Added wholesale storefront entry with catalog/product browsing, a wholesale cart, and a quick-order flow.
- **Bug Fixes**
  - Improved robustness when monetary totals or item/pricing fields are missing, preventing parsing/rendering issues.
  - Strengthened cart/checkout behavior to stay correctly separated between standard and wholesale flows.
- **Documentation**
  - Updated `.env.example` and README with wholesale enablement and channel-scoped publishable key guidance.

### CODE REVIEW

**Comment by damianlegawiec** [comment]
Addressed all four Bugbot findings in 6c3e73d — all four were real:

**Logout skips wholesale cart (High)** — Fixed. `logout()` now calls a new `clearAllCartCookies()` and invalidates every surface's `cart` cache tag (via a new `SURFACES` constant), so the wholesale cart cookie no longer survives sign-out. The login-flow guest-cart-association path stays DTC-scoped intentionally (the wholesale channel 401s guests, so there's no wholesale guest cart there).

**Gate children run before auth (High)** — Fixed. `WholesaleGate` now takes its approved-buyer content as a thunk (`children: () => ReactNode`), invoked only in the approved branch. The gated product fetch can no longer run for a guest, so guests always get the sign-in wall, never a 404. All four gated pages updated.

**Wholesale cart drawer never opens (Medium)** — Fixed. Mounted `<CartDrawer />` inside the wholesale `<CartProvider>` in the gate, so it reads the same context that wholesale add-to-cart / quick-order write to. The root layout's drawer stays bound to the outer DTC provider.

**Checkout surface needs cookies (High)** — Fixed, and I'd initially under-weighted this one — the offsite-return path's own comment admits cookies may be cleared. `confirmPaymentAndCompleteCart` now resolves the surface via a new `resolveSurfaceForCartVerified()` that confirms against the cart's `channel_id` when the cookie doesn't identify it, and threads that verified surface through completion so nothing re-resolves from a dropped cookie. In-session actions keep the cheap cookie-only path (cookie always present); DTC-only storefronts short-circuit with no extra fetch.

tsc + biome clean, 115/115 tests pass (updated the affected mocks).

**Comment by damianlegawiec** [comment]
Addressed the CodeRabbit findings in d49259a:

**express-checkout — gate discount/tax on item_total (Minor):** Fixed. `buildLineItems` now returns no line items at all when `item_total` is null, so the Express Checkout Element falls back to the total-only amount rather than emitting orphan Discount/Tax lines against a missing subtotal.

**Duplicate isPoisonedDtcCartId (Major):** Fixed. Consolidated into a single exported helper in `cookies.ts`; `cart.ts` and the `requireCartId` path both import it. Kept the correct `surface !== "dtc"` semantics (poisoning is specifically a DTC-cookie-holds-wholesale-id problem) and removed the divergent inverted copy.

**README wholesale docs link (Minor):** Keeping the URL as-is. It's the correct eventual path and identical in form to the sibling `spreecommerce.org/docs/developer/storefront/nextjs/...` links already in the README (e.g. quickstart). It 404s only because the docs page — added in the `spree/spree` PR — hasn't published yet; it resolves once that deploys. Changing to a different/relative URL would be wrong.

Separately, the four earlier Bugbot findings (logout cart clear, gate children eval order, wholesale cart drawer, checkout surface after offsite redirect) were fixed in 6c3e73d.

tsc + biome clean, 115/115 tests pass.

File: src/lib/data/checkout.ts [hunk 1]
```diff
@@ -3,30 +3,94 @@
...
+  // failure (not a wholesale cart, gated 401, network) fall back to DTC.
+  return withFallback(async () => {
+    const [cart, channel] = await Promise.all([
+      getCart(cartId, "wholesale"),
+      getWholesaleChannel(),
+    ]);
+    return cart && channel && cart.channel_id === channel.id
+      ? "wholesale"
+      : "dtc";
+  }, "dtc");
```
[the remark this answers is not in the document]
Comment (damianlegawiec) [reply src/lib/data/checkout.ts]: Fixed in fa20123 (logic) / e6e2c26. `resolveSurfaceForCartVerified` now returns a distinct `"unverified"` when the wholesale lookup can't complete, and `confirmPaymentAndCompleteCart` fails closed on it (returns an error asking the buyer to retry) rather than defaulting to DTC. Only a positive channel-match/mismatch decides the surface. Covered by `resolve-surface.test.ts`.

File: src/lib/data/customer.ts [hunk 2]
```diff
@@ -188,9 +194,12 @@ export async function logout(): Promise<void> {
...
   await clearAccessToken();
   await clearRefreshToken();
-  await clearCartCookies();
+  // Clear every surface's cart — the wholesale cart lives in its own cookie
+  // pair and cache tag, so a DTC-only clear would leave it behind for the
+  // next session.
+  await clearAllCartCookies();
   updateTag("customer");
-  updateTag("cart");
+  for (const surface of SURFACES) updateTag(`cart${cacheTagSuffix(surface)}`);
```
[the remark this answers is not in the document]
Comment (damianlegawiec) [reply src/lib/data/customer.ts]: Fixed in e6e2c26. `logout` now invalidates `checkout${cacheTagSuffix(surface)}` alongside the cart tag for every surface; the logout test asserts both surfaces' cart + checkout tags.

File: src/lib/utils/express-checkout.ts [hunk 3]
```diff
@@ -56,15 +56,23 @@ export function buildLineItems(order: Cart) {
...
-  const itemTotal = toCents(order.item_total, currency);
-  items.push({ name: "Subtotal", amount: itemTotal });
+  // The subtotal anchors the breakdown. When it's unknown — money fields are
+  // nullable for prices-hidden carts — return no line items at all so the
+  // Express Checkout Element falls back to the total-only amount, rather than
+  // emitting orphan Discount/Tax lines against a missing subtotal.
+  if (order.item_total == null) return items;
 
-  const promoTotal = toCents(order.discount_total, currency);
+  items.push({ name: "Subtotal", amount: toCents(order.item_total, currency) });
```
[the remark this answers is not in the document]
Comment (damianlegawiec) [reply src/lib/utils/express-checkout.ts:78]: Fixed in e6e2c26. Added `hasPayableTotal()`; `ExpressCheckoutButton` now reports unavailable and returns null when the cart has no positive payable total, so a prices-hidden (null-total) cart never mounts Stripe Elements with `amount: 0`. Covered by `express-checkout.test.ts`.

File: src/app/[country]/[locale]/(wholesale)/wholesale/_components/WholesaleGuestBrowse.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+  return (
+    <HiddenPricingProvider value={{ signInHref }}>
+      <WholesaleHeader
+        basePath={basePath}
+        authenticated={false}
+        signInHref={signInHref}
+      />
+      <main>{children}</main>
+    </HiddenPricingProvider>
```
[the remark this answers is not in the document]
Comment (damianlegawiec) [reply src/app/[country]/[locale]/(wholesale)/wholesale/_components/WholesaleGuestBrowse.tsx:42]: Fixed in fa20123. `WholesaleGate` now gates guest browsing behind an `allowGuestBrowse` flag set only on the catalog + PDP; cart and quick-order wall guests off (no DTC-bound `useCart()`), and the guest header no longer exposes Quick Order.

File: src/components/products/HiddenPricePrompt.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      <Lock className="h-3.5 w-3.5" />
+      {t("hiddenPrice.signInForPricing")}
+    </Link>
+  );
+}
+
+/** Whether a null/absent price should render the sign-in prompt in this context. */
+export function useIsPriceHidden(): boolean {
+  return useHiddenPricing() !== null;
+}
```
[the remark this answers is not in the document]
Comment (damianlegawiec) [reply src/components/products/HiddenPricePrompt.tsx]: Fixed in fa20123 — removed the unused `useIsPriceHidden` helper.

File: README.md [hunk 6]
```diff
@@ -129,6 +131,14 @@ SPREE_PUBLISHABLE_KEY=your_publishable_api_key_here
 
 > **Privacy note:** PII collection is disabled by default. Only set `SENTRY_SEND_DEFAULT_PII` / `NEXT_PUBLIC_SENTRY_SEND_DEFAULT_PII` to `true` if you have appropriate user consent or a privacy policy covering this data.
 
+#### Wholesale B2B portal (opt-in addon)
+
+The storefront ships an optional gated wholesale portal at `/wholesale`, a B2B surface that runs on a separate Spree channel while sharing the same storefront and backend as DTC.
+
+`SPREE_WHOLESALE_CHANNEL` is the enable switch and has **no default**. Leave it unset and the storefront is DTC-only: the wholesale nav link, footer link, homepage section, and the `/wholesale` routes are all hidden — the routes return 404. Set it to the code of a gated channel on your Spree backend (e.g. `wholesale`) to turn the portal on. `SPREE_WHOLESALE_PUBLISHABLE_KEY` is optional and only needed to bind a channel-scoped publishable key; without it the surface reuses `SPREE_PUBLISHABLE_KEY` and selects the channel via the `X-Spree-Channel` header.
+
+See the [wholesale portal guide](https://spreecommerce.org/docs/developer/storefront/nextjs/wholesale) for how channel switching works end to end.
```
[the remark this answers is not in the document]
Comment (damianlegawiec) [reply README.md:140]: Skipping: the guide link is a forward-reference to the wholesale docs page added in spree/spree (docs/developer/storefront/nextjs/wholesale.mdx); it resolves once that docs build publishes. Keeping the canonical URL rather than pointing at a temporary path.
